hashing/md5.cpp

```diff
@@ -0,0 +1,373 @@
+        sig[i + 12] = (d0 >> (8 * i)) & 0xFF;
+    }
+
+    return sig;
+}
+/**
+ * @brief Converts the string to bytestring and calls the main algorithm
+ * @param message Plain character message to hash
+ * @return void* Pointer to the MD5 signature
+ */
+void* hash(const std::string& message) {
+    return hash_bs(&message[0], message.size());
+}
+}  // namespace md5
+}  // namespace hashing
+
+/**
+ * @brief Self-test implementations of well-known MD5 hashes
+ * @returns void
+ */
+static void test() {
+    // Hashes empty string and stores signature
+    void* sig = hashing::md5::hash("");
+    std::cout << "Hashing empty string" << std::endl;
+    // Prints signature hex representation
+    std::cout << hashing::md5::sig2hex(sig) << std::endl << std::endl;
+    // Test with cassert whether sig is correct from the expected value
+    assert(hashing::md5::sig2hex(sig).compare(
+               "d41d8cd98f00b204e9800998ecf8427e") == 0);
+
+    // Hashes "The quick brown fox jumps over the lazy dog" and stores signature
+    void* sig2 =
+        hashing::md5::hash("The quick brown fox jumps over the lazy dog");
+    std::cout << "Hashing The quick brown fox jumps over the lazy dog"
+              << std::endl;
+    // Prints signature hex representation
+    std::cout << hashing::md5::sig2hex(sig2) << std::endl << std::endl;
+    // Test with cassert whether sig is correct from the expected value
+    assert(hashing::md5::sig2hex(sig2).compare(
+               "9e107d9d372bb6826bd81d3542a419d6") == 0);
+
+    // Hashes "The quick brown fox jumps over the lazy dog." (notice the
+    // additional period) and stores signature
+    void* sig3 =
+        hashing::md5::hash("The quick brown fox jumps over the lazy dog.");
+    std::cout << "Hashing "
+                 "The quick brown fox jumps over the lazy dog."
+              << std::endl;
+    // Prints signature hex representation
+    std::cout << hashing::md5::sig2hex(sig3) << std::endl << std::endl;
+    // Test with cassert whether sig is correct from the expected value
+    assert(hashing::md5::sig2hex(sig3).compare(
+               "e4d909c290d0fb1ca068ffaddf22cbd0") == 0);
+
+    // Hashes "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
+    // and stores signature
+    void* sig4 = hashing::md5::hash(
+        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789");
+    std::cout
+        << "Hashing "
+           "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
+        << std::endl;
+    // Prints signature hex representation
+    std::cout << hashing::md5::sig2hex(sig4) << std::endl << std::endl;
+    // Test with cassert whether sig is correct from the expected value
+    assert(hashing::md5::sig2hex(sig4).compare(
+               "d174ab98d277d9f5a5611c2c9f419d9f") == 0);
+}
+
+/**
+ * @brief Puts user in a loop where inputs can be given and MD5 hash will be
+ * computed and printed
+ * @returns void
+ */
+static void interactive() {
```

Comment:
I like this! It works very well, but I would recommend adding an option for the user to choose if they want to stay inside the interactive mode. I was trying to find a way to quit but then realized it was an infinite loop. 😅